### ISSUE TITLE
"Download This Book" Select is inaccessible

### ISSUE THREAD

**Issue by avaaec** [pressbooks/pressbooks#1790]
### Prerequisites

* [X] Can you reproduce the behavior you wish to address with [all other plugins deactivated](http://codex.wordpress.org/Multisite_Network_Administration#Plugins)?
* [X] Are you running the [latest version of Pressbooks](http://github.com/pressbooks/pressbooks/releases/latest/) and the [latest version of WordPress](http://codex.wordpress.org/Upgrading_WordPress)?
* [x] Are you experiencing this issue with one of our built-in root or book themes?
* [X] Did you [perform a cursory search](https://github.com/issues?q=+is%3Aissue+repo%3Apressbooks%2Fpressbooks) to see if your bug or enhancement is already listed?

### Steps to Reproduce

1. Go to a book that can be downloaded ([example](https://uw.pressbooks.pub/bookclubs/))
2. Navigate to the "Download This Book" dropdown select widget
3. Use the keyboard to open the menu
4. Try to use the arrow keys to navigate through the options
5. Try to use Escape to close the menu
6. Try to use Tab to move past the menu

**Expected behavior:** 
It should follow the [aria spec](https://www.w3.org/TR/wai-aria-practices-1.1/#Listbox) for a dropdown listbox. This can be done trivially with a HTML5 select element.
**Actual behavior:** 
The current implementation is very non standard and does not follow most of the aria keyboard navigation criteria. Additionally it is buggy to use with a mouse, sometimes staying open when the user clicks somewhere else on the page.

**Comment by sonofborge** [comment]
@connerbw, I was looking into this ticket a bit last night, and at first it looked like there were two approaches, but I'm thinking there might be just one.  Also worth mentioning that I think all of the code changes can be done in pressbooks-book, and no changes need to be made in pressbooks proper (unless I've missed something).

It would be an arbitrary fix to convert the existing ul/li elements to select/option elements and tweak the styles and js accordingly, but the part that makes this tricky is the analytics tracking filter that's being defined in the foreach loop (for each item in the dropdown list).  Unfortunately, it won't work to wrap the option element with an anchor, nor can you set the element to listen to onclick or onchange events (you have to do that on the select element).  I went down that path for a minute and things started to get dicey, but definitely open to suggestions and thoughts if I'm missing something.

The other way is to keep the elements mostly as is and set an appropriate role attribute on each, and then write a fair bit of javascript to handle keyup and click behavior on the listbox.  It might make a lot of sense to change up how the javascript is currently written, and maybe break it out into a separate `Listbox` class, but I'm not sure what the impact of that would be where that code is used elsewhere.

I've started to get pretty far in with the latter approach, and would be happy to submit a PR if it sounds like this is a good way to go.

**Comment by SteelWagstaff** [comment]
Brian--thanks for looking into this! We'll be discussing on Monday in a bug backlog refinement meeting and should have more thoughts for you soon(ish).

**Comment by sonofborge** [comment]
Sounds good @SteelWagstaff.  Here's WIP so far on this.  PR isn't up yet as mentioned in my comments:

https://github.com/sonofborge/pressbooks-book/commit/c9d6e79c5bd40f114b62b04e95555092e879e251

**Comment by avaaec** [comment]
This is a UW ATS priority 0 issue

**Comment by avaaec** [comment]
Also @SteelWagstaff Thank you for making a UW tag! We really appreciate your efforts!

**Comment by SteelWagstaff** [comment]
At minimum, solution should:
- [ ] add `role="listbox"` to ul in https://github.com/pressbooks/pressbooks-book/blob/108f5d4d0324adc5436a16cd2edac0f70646269a/partials/content-cover-book-header.php#L55
- [ ] add `role="option" to each li https://github.com/pressbooks/pressbooks-book/blob/108f5d4d0324adc5436a16cd2edac0f70646269a/partials/content-cover-book-header.php#L75
- [ ] fix closing `<ul>` tag by changing to `</ul>`: https://github.com/pressbooks/pressbooks-book/blob/108f5d4d0324adc5436a16cd2edac0f70646269a/partials/content-cover-book-header.php#L84

**Comment by sonofborge** [comment]
@SteelWagstaff, that's a good minimum solution you've laid out.  +1

My branch (linked above in the comments) might be a good resource for anyone taking the next step to handle the issues I mentioned.  I'd be happy to submit a PR that at least handles the minimum you outlined.  Let me know!  Cheers!  :)

**Comment by greatislander** [comment]
Hi all! I may be completely off base but this doesn't seem to fit the criteria for a listbox as each item needs to function as a link, not as a static select (as @sonofborge identified, you can't wrap an `<option>` in an anchor).

What I did with the dropdown appears to fit this WAI example for fly-out menus: https://www.w3.org/WAI/tutorials/menus/flyout/#use-parent-as-toggle

I'm happy to be corrected, but I wonder if there are improvements might be made to better align this with that existing pattern rather than rewriting the control?

**Comment by SteelWagstaff** [comment]
@greatislander Thanks for chiming in. Having looked at this fly-out menu example and what exists for the download this book option, I understand better what's happening and how to navigate it as a keyboard user. 

The only thing that I'd note about the current implementation is that the WAI example suggests 
> the submenu is opened by a script when the user activates the top-level item and closed when the focus leaves the submenu

As @avaaec noted, the menu remains open even when focus leaves the menu. Here's a video example I recorded using keyboard navigation to open/close the menu (with the enter key) and then tabbing through the options: https://youtu.be/oS-7KgRIYPE. 

You'll notice that the menu remains open even after tabbing out of the menu and into the next available selectable link on the page. Perhaps this should be improved?

**Comment by greatislander** [comment]
@SteelWagstaff That could definitely be implemented within the existing pattern; furthermore, using `esc` to dismiss the menu could also be added.

**Comment by SteelWagstaff** [comment]
Thanks for discussion on this everyone. Rather than redoing the existing pattern in favor of implementing this feature as a listbox, I propose that we improve the accessibility of this feature by 
1. closing the menu when the user tabs out of the menu and 
2. allowing the user to dismiss the menu by pressing the `esc` key.

**Comment by sonofborge** [comment]
Good call @greatislander and @SteelWagstaff.  +1

**Comment by avaaec** [comment]
Hi again! When I initially reported this issue I didn't notice that the items in the flyout were links! My apologies. In light of this I don't think a select / popup listbox is the right choice. The accessibility fixes proposed look adequate, anther option could be to use an aria menu:
https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-links.html

**Comment by greatislander** [comment]
@sonofborge Still up for putting a patch together for the two points in @SteelWagstaff's comment? If not I might have time to do something at the end of the week!

**Comment by sonofborge** [comment]
@greatislander , yeah I can take a stab at this early next week.  Cheers!

**Comment by sonofborge** [comment]
Alright, a PR is open and I've commented as thoroughly as I can think to right now.  Please don't hesitate to ask questions or point out things I may have missed (or something that doesn't look right).  Also, happy to continue the conversation either here or in the PR.  Cheers!

**Comment by cagp-dev-mtl** [comment]
Hello @sonofborge , thank you very much for this contribution, merging this PR now. :+1:

**Comment by SteelWagstaff** [comment]
@cagp-dev-mtl has dev branch with changes added by https://github.com/pressbooks/pressbooks-book/pull/641 been deployed to integrations and/or textopress? I went to go test today but wasn't seeing changes as expected ...

**Comment by SteelWagstaff** [comment]
Looking good to me now. @mcgratay ready for final testing from you. We can discuss in tomorrow's scrum.

**Comment by mcgratay** [comment]
This looks good to go. There are some lingering issues with the Safari browser we'll address in another ticket.

**Comment by connermwallace** [comment]
University of Washington Accessibility Technology Services has inspected this issue and we have verified that it has been fixed. Version: 5.3.2.
